src/storage/data_table.cpp

```diff
@@ -246,6 +258,251 @@ bool DataTable::Delete(transaction::TransactionContext *const txn, const TupleSl
+  len = StorageUtil::PadUpToSize(ARROW_ALIGNMENT, len);
+  buffers->emplace_back(*offset, len);
+  *offset += len;
+}
+
+void DataTable::WriteSchemaMessage(std::ofstream &outfile, std::unordered_map<col_id_t, int64_t> *dictionary_ids,
+                                   flatbuffers::FlatBufferBuilder *flatbuf_builder) {
+  RawBlock *block = blocks_.front();
+  const BlockLayout &layout = accessor_.GetBlockLayout();
+  ArrowBlockMetadata &metadata = accessor_.GetArrowBlockMetadata(block);
+  std::vector<flatbuffers::Offset<flatbuf::Field>> fields;
+  int64_t dictionary_id = 0;
+  for (col_id_t col_id : layout.AllColumns()) {
+    ArrowColumnInfo &col_info = metadata.GetColumnInfo(layout, col_id);
+    // TODO(Yuze): Change column name when we have the information, which we may need from upper layers.
+    auto name = flatbuf_builder->CreateString("Col" + std::to_string(!col_id));
+    flatbuf::Type type;
+    flatbuffers::Offset<void> type_offset;
+    flatbuffers::Offset<flatbuf::DictionaryEncoding> dictionary = 0;
+    if (!layout.IsVarlen(col_id) || col_info.Type() == ArrowColumnType::FIXED_LENGTH) {
+      uint8_t byte_width = accessor_.GetBlockLayout().AttrSize(col_id);
+      type = flatbuf::Type_FixedSizeBinary;
+      type_offset = flatbuf::CreateFixedSizeBinary(*flatbuf_builder, byte_width).Union();
+    } else {
+      switch (col_info.Type()) {
+        case ArrowColumnType::DICTIONARY_COMPRESSED:
+          dictionary = flatbuf::CreateDictionaryEncoding(
+              *flatbuf_builder, dictionary_id, flatbuf::CreateInt(*flatbuf_builder, 8 * sizeof(uint64_t), true), false);
+          dictionary_ids->emplace(col_id, dictionary_id++);
+          TERRIER_FALLTHROUGH;
+        case ArrowColumnType::GATHERED_VARLEN:
+          type = flatbuf::Type_LargeBinary;
+          type_offset = flatbuf::CreateLargeBinary(*flatbuf_builder).Union();
+          break;
+        default:
+          throw std::runtime_error("unexpected control flow");
+      }
+    }
+
+    // Apache Arrow supports nested logical types. For example, for type List<Int64>, the parent type is List,
+    // and its children type is Int64. Another example, for type List<List<Int64>>, List<Int64> is the children
+    // of the outer List, and Int64 is the children of the inner List. Since we don't have nested types, we use
+    // an empty array as fake children.
+    std::vector<flatbuffers::Offset<org::apache::arrow::flatbuf::Field>> fake_children;
+    fields.emplace_back(flatbuf::CreateField(*flatbuf_builder, name, true, type, type_offset, dictionary,
+                                             flatbuf_builder->CreateVector(fake_children)));
+  }
+  auto schema =
+      flatbuf::CreateSchema(*flatbuf_builder, flatbuf::Endianness_Little, flatbuf_builder->CreateVector(fields));
+  auto message = flatbuf::CreateMessage(*flatbuf_builder, flatbuf::MetadataVersion_V4, flatbuf::MessageHeader_Schema,
```

Comment:
Apache Arrow uses MetadataVersion_V4 when they write an arrow-format message. When they read a message, they support all the possible versions (V1-V4):

[https://github.com/apache/arrow/blob/4c63bef8de079d4e6e559adab822d0f44dc8645d/cpp/src/arrow/ipc/metadata_internal.h#L65](https://github.com/apache/arrow/blob/4c63bef8de079d4e6e559adab822d0f44dc8645d/cpp/src/arrow/ipc/metadata_internal.h#L65)

[https://github.com/apache/arrow/blob/4c63bef8de079d4e6e559adab822d0f44dc8645d/cpp/src/arrow/ipc/metadata_internal.cc#L851](https://github.com/apache/arrow/blob/4c63bef8de079d4e6e559adab822d0f44dc8645d/cpp/src/arrow/ipc/metadata_internal.cc#L851)

Since we only write in our system, we may use some other versions. But use V4 seems to be consistent with Apache Arrow. As long as they guarantee "read" is always compatible with all the versions, we don't need to change this in the future because we only do "write" things.